Client/Extensions/PaymentRequestExtension.swift

```diff
@@ -0,0 +1,146 @@
+
+class PaymentRequestExtension: NSObject {
+    fileprivate weak var tab: Tab?
+    fileprivate var response = ""
+    fileprivate var token: String
+    
+    init(tab: Tab) {
+        token = UserScriptManager.securityToken.uuidString.replacingOccurrences(of: "-", with: "", options: .literal)
+        self.tab = tab
+    }
+}
+
+extension PaymentRequestExtension: TabContentScript {
+    static func name() -> String {
+        return "PaymentRequest"
+    }
+    
+    func scriptMessageHandlerName() -> String? {
+        return "PaymentRequest"
+    }
+    
+    func sendPaymentRequestError(errorName: String, errorMessage: String) {
+        ensureMainThread {
+            self.tab?.webView?.evaluateJavaScript("PaymentRequestCallback\(self.token).paymentreq_postCreate('', '\(errorName)', '\(errorMessage)')", completionHandler: { _, error in
+                    if error != nil {
+                        log.error(error)
+                    }
+                })
+        }
+    }
+    
+    func userContentController(_ userContentController: WKUserContentController, didReceiveScriptMessage message: WKScriptMessage) {
+        if message.name == "PaymentRequest", let body = message.body as? NSDictionary {
+            do {
+                let messageData = try JSONSerialization.data(withJSONObject: body, options: [])
+                let body = try JSONDecoder().decode(PaymentRequestBodyParser.self, from: messageData)
+                if body.name == "payment-request-show" {
+                    popup.initPaymentUI()
+                    
+                    guard let detailsData = body.details.data(using: String.Encoding.utf8), let supportedInstrumentsData = body.supportedInstruments.data(using: String.Encoding.utf8) else {
+                        log.error("Error parsing data")
+                        return
+                    }
+                    
+                    let details = try JSONDecoder().decode(PaymentRequestDetailsParser.self, from: detailsData)
+                    
+                    let supportedInstruments =  try JSONDecoder().decode([PaymentRequestSupportedInstrumentsParser].self, from: supportedInstrumentsData)
+                    
+                    guard supportedInstruments.contains(where: {$0.supportedMethods == "bat"}) else {
+                        sendPaymentRequestError(errorName: Strings.notSupportedErrorName, errorMessage: Strings.unsupportedInstrumentMessage)
+                        return
+                    }
+                    
+                    for item in details.displayItems {
+                        popup.addDisplayItemLabel(message: "\(item.label): \(item.amount.value) \(item.amount.currency)\n")
+                        log.info(item.label)
+                    }
+                    
+                    popup.addTotalLabel(message: details.total.label + ":  " + details.total.amount.value + " " + details.total.amount.currency + "\n")
```

Comment:
Swift has inline string evaluation similar to JS "`"

```suggestion
                    popup.addTotalLabel(message: "\(details.total.label):  \(details.total.amount.value) \(details.total.amount.currency)\n")
```